# 01. Importing Libraries

In [5]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [6]:
# Import the data set into Jupyter as a new dataframe
path = r'/Users/ChuahLH/Desktop/Documents/Instacart Basket Analysis'

In [7]:
# Import the orders_products_merged dataframe
orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))

# 03. Groupby & Aggregating

In [8]:
# Investigate the df_orders_products_merged
c.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0        0.0           195         Soda        77              7     9.0   
1        1.0           195         Soda        77              7     9.0   
2        1.0           195         Soda        77              7     9.0   
3        1.0           195         Soda        77              7     9.0   
4        1.0           195         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product      Least busy      Least busy days   
2   both  Mid-range product      Least busy      Least busy days   
3   both  Mid-range product      Least busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

In [9]:
df = orders_products_merged

In [12]:
df.shape

(32406041, 20)

In [14]:
df.head(10)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   
5             5   3367565        1             6                   2   
6             6    550135        1             7                   1   
7             7   3108588        1             8                   1   
8             8   2295261        1             9                   1   
9             9   2550362        1            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   
5                  7                   19.0       196.0                1.0   
6                  9                   20.0       196.0                1.0   
7                 14                   14.0       196.0                2.0   
8                 16                    0.0       196.0                4.0   
9                  8                   30.0       196.0                1.0   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0        0.0           195         Soda        77              7     9.0   
1        1.0           195         Soda        77              7     9.0   
2        1.0           195         Soda        77              7     9.0   
3        1.0           195         Soda        77              7     9.0   
4        1.0           195         Soda        77              7     9.0   
5        1.0           195         Soda        77              7     9.0   
6        1.0           195         Soda        77              7     9.0   
7        1.0           195         Soda        77              7     9.0   
8        1.0           195         Soda        77              7     9.0   
9        1.0           195         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product      Least busy      Least busy days   
2   both  Mid-range product      Least busy      Least busy days   
3   both  Mid-range product      Least busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   
5   both  Mid-range product  Regularly busy  Regularly busy days   
6   both  Mid-range product     Busiest day         Busiest days   
7   both  Mid-range product     Busiest day         Busiest days   
8   both  Mid-range product     Busiest day         Busiest days   
9   both  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  
5        Average orders  
6           Most orders  
7           Most orders  
8           Most orders  
9        Average orders

In [15]:
# groupby product name
df.groupby('product_name')

In [16]:
# Performing single aggregation
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.213779
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.171095
12               15.887671
13               16.583536
14               16.771241
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [17]:
# Performing multiple aggregation
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.213779   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.171095   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.771241   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [18]:
# Aggreting data with transform()
orders_products_merged['max_order'] = orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [19]:
orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
..           ...       ...      ...           ...                 ...   
95          5895   3226575      360             1                   5   
96          6137   1469869      377             3                   5   
97          6259   1927023      387             2                   4   
98          6740    858092      420             4                   1   
99          6741    119314      420             5                   3   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN       196.0                1.0   
1                   7                   15.0       196.0                1.0   
2                  12                   21.0       196.0                1.0   
3                   7                   29.0       196.0                1.0   
4                  15                   28.0       196.0                1.0   
..                ...                    ...         ...                ...   
95                 12                    NaN       196.0                1.0   
96                 17                    3.0       196.0                9.0   
97                 10                   22.0       196.0                3.0   
98                 19                   30.0       196.0                2.0   
99                 19                   16.0       196.0                2.0   

    reordered  ...  product_name aisle_id  department_id  prices  _merge  \
0         0.0  ...          Soda       77              7     9.0    both   
1         1.0  ...          Soda       77              7     9.0    both   
2         1.0  ...          Soda       77              7     9.0    both   
3         1.0  ...          Soda       77              7     9.0    both   
4         1.0  ...          Soda       77              7     9.0    both   
..        ...  ...           ...      ...            ...     ...     ...   
95        0.0  ...          Soda       77              7     9.0    both   
96        0.0  ...          Soda       77              7     9.0    both   
97        0.0  ...          Soda       77              7     9.0    both   
98        0.0  ...          Soda       77              7     9.0    both   
99        1.0  ...          Soda       77              7     9.0    both   

      price_range_loc     busiest_day         busiest_days  \
0   Mid-range product  Regularly busy  Regularly busy days   
1   Mid-range product      Least busy      Least busy days   
2   Mid-range product      Least busy      Least busy days   
3   Mid-range product      Least busy      Least busy days   
4   Mid-range product      Least busy      Least busy days   
..                ...             ...                  ...   
95  Mid-range product  Regularly busy  Regularly busy days   
96  Mid-range product  Regularly busy  Regularly busy days   
97  Mid-range product      Least busy      Least busy days   
98  Mid-range product     Busiest day         Busiest days   
99  Mid-range product      Least busy      Least busy days   

   busiest_period_of_day max_order  
0         Average orders        10  
1         Average orders        10  
2            Most orders        10  
3         Average orders        10  
4            Most orders        10  
..                   ...       ...  
95           Most orders         3  
96        Average orders         3  
97           Most orders         8  
98        Average orders        22  
99        Average orders        22  

[100 rows x 21 columns]

In [20]:
pd.options.display.max_rows = None

In [21]:
# Deriving Columns with loc()
orders_products_merged.loc[orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [22]:
orders_products_merged.loc[(orders_products_merged['max_order'] <= 40) & (orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [23]:
orders_products_merged.loc[orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [24]:
orders_products_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15877395
Loyal customer      10284396
New customer         6244250
Name: loyalty_flag, dtype: int64

In [25]:
orders_products_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  ...  aisle_id department_id  prices  _merge    price_range_loc  \
0        0.0  ...        77             7     9.0    both  Mid-range product   
1        1.0  ...        77             7     9.0    both  Mid-range product   
2        1.0  ...        77             7     9.0    both  Mid-range product   
3        1.0  ...        77             7     9.0    both  Mid-range product   
4        1.0  ...        77             7     9.0    both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy days        Average orders        10   
1      Least busy      Least busy days        Average orders        10   
2      Least busy      Least busy days           Most orders        10   
3      Least busy      Least busy days        Average orders        10   
4      Least busy      Least busy days           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 22 columns]

In [26]:
orders_products_merged[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# 04. Groupby & Aggregating (1 Million Rows)

In [27]:
# Only work with a subset of the dataframe for the first one million rows
df = orders_products_merged[:1000000]

In [28]:
df.shape

(1000000, 22)

In [29]:
df.head(10)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   
5             5   3367565        1             6                   2   
6             6    550135        1             7                   1   
7             7   3108588        1             8                   1   
8             8   2295261        1             9                   1   
9             9   2550362        1            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   
5                  7                   19.0       196.0                1.0   
6                  9                   20.0       196.0                1.0   
7                 14                   14.0       196.0                2.0   
8                 16                    0.0       196.0                4.0   
9                  8                   30.0       196.0                1.0   

   reordered  ...  aisle_id department_id  prices  _merge    price_range_loc  \
0        0.0  ...        77             7     9.0    both  Mid-range product   
1        1.0  ...        77             7     9.0    both  Mid-range product   
2        1.0  ...        77             7     9.0    both  Mid-range product   
3        1.0  ...        77             7     9.0    both  Mid-range product   
4        1.0  ...        77             7     9.0    both  Mid-range product   
5        1.0  ...        77             7     9.0    both  Mid-range product   
6        1.0  ...        77             7     9.0    both  Mid-range product   
7        1.0  ...        77             7     9.0    both  Mid-range product   
8        1.0  ...        77             7     9.0    both  Mid-range product   
9        1.0  ...        77             7     9.0    both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy days        Average orders        10   
1      Least busy      Least busy days        Average orders        10   
2      Least busy      Least busy days           Most orders        10   
3      Least busy      Least busy days        Average orders        10   
4      Least busy      Least busy days           Most orders        10   
5  Regularly busy  Regularly busy days        Average orders        10   
6     Busiest day         Busiest days           Most orders        10   
7     Busiest day         Busiest days           Most orders        10   
8     Busiest day         Busiest days           Most orders        10   
9      Least busy      Least busy days        Average orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  
5  New customer  
6  New customer  
7  New customer  
8  New customer  
9  New customer  

[10 rows x 22 columns]

In [30]:
# groupby product name
df.groupby('product_name')

In [31]:
# Performing single aggregation
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [32]:
# Performing multiple aggregation
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [33]:
# Aggreting data with transform()
orders_products_merged['max_order'] = orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [34]:
orders_products_merged.head(15)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
5              5   3367565        1             6                   2   
6              6    550135        1             7                   1   
7              7   3108588        1             8                   1   
8              8   2295261        1             9                   1   
9              9   2550362        1            10                   4   
10           144   2968173       15            15                   1   
11           146   1870022       15            17                   2   
12           147   1911383       15            18                   2   
13           150   2715276       15            21                   1   
14           151    487368       15            22                   1   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN       196.0                1.0   
1                   7                   15.0       196.0                1.0   
2                  12                   21.0       196.0                1.0   
3                   7                   29.0       196.0                1.0   
4                  15                   28.0       196.0                1.0   
5                   7                   19.0       196.0                1.0   
6                   9                   20.0       196.0                1.0   
7                  14                   14.0       196.0                2.0   
8                  16                    0.0       196.0                4.0   
9                   8                   30.0       196.0                1.0   
10                  9                    7.0       196.0                2.0   
11                 16                    8.0       196.0                6.0   
12                 11                    7.0       196.0                1.0   
13                  9                    7.0       196.0                1.0   
14                 10                   14.0       196.0                1.0   

    reordered  ...  aisle_id department_id  prices  _merge    price_range_loc  \
0         0.0  ...        77             7     9.0    both  Mid-range product   
1         1.0  ...        77             7     9.0    both  Mid-range product   
2         1.0  ...        77             7     9.0    both  Mid-range product   
3         1.0  ...        77             7     9.0    both  Mid-range product   
4         1.0  ...        77             7     9.0    both  Mid-range product   
5         1.0  ...        77             7     9.0    both  Mid-range product   
6         1.0  ...        77             7     9.0    both  Mid-range product   
7         1.0  ...        77             7     9.0    both  Mid-range product   
8         1.0  ...        77             7     9.0    both  Mid-range product   
9         1.0  ...        77             7     9.0    both  Mid-range product   
10        0.0  ...        77             7     9.0    both  Mid-range product   
11        1.0  ...        77             7     9.0    both  Mid-range product   
12        1.0  ...        77             7     9.0    both  Mid-range product   
13        1.0  ...        77             7     9.0    both  Mid-range product   
14        1.0  ...        77             7     9.0    both  Mid-range product   

       busiest_day         busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy days        Average orders        10   
1       Least busy      Least busy days        Average orders        10   
2       Least busy      Least busy days         

In [35]:
orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
5              5   3367565        1             6                   2   
6              6    550135        1             7                   1   
7              7   3108588        1             8                   1   
8              8   2295261        1             9                   1   
9              9   2550362        1            10                   4   
10           144   2968173       15            15                   1   
11           146   1870022       15            17                   2   
12           147   1911383       15            18                   2   
13           150   2715276       15            21                   1   
14           151    487368       15            22                   1   
15           209   2293453       19             2                   5   
16           212   1973799       19             5                   6   
17           214    532817       19             7                   4   
18           232   1573906       21            10                   3   
19           458   1593000       31            10                   3   
20           465   2231262       31            17                   3   
21           617   2580647       43             6                   4   
22           620   2187180       43             9                   4   
23           751   2497897       52             1                   1   
24           752   1318871       52             2                   1   
25           753   1261384       52             3                   1   
26           755    580568       52             5                   2   
27           757   2428073       52             7                   2   
28           759    342306       52             9                   1   
29           760   1520353       52            10                   2   
30           761   2542086       52            11                   2   
31           762    944694       52            12                   1   
32           765   1268191       52            15                   2   
33           767   1122089       52            17                   2   
34           768   1498922       52            18                   2   
35           769    180919       52            19                   3   
36           771   2695875       52            21                   5   
37          1022   3205960       67             1                   3   
38          1023    970922       67             2                   1   
39          1024    596585       67             3                   1   
40          1025   2037087       67             4                   2   
41          1026   2474111       67             5                   3   
42          1027   1568481       67             6                   3   
43          1029   2048527       67             8                   2   
44          1030   2389031       67             9                   2   
45          1031   2378562       67            10                   1   
46          1032   1515698       67            11                   1   
47          1033   1928050       67            12                   2   
48          1034   1949274       67            13                   1   
49          1035    551869       67            14                   1   
50          1037    768572       67            16                   4   
51          1039    150528       67            18                   5   
52          1040     29758       67            19                   0   
53          1042   1558976       67            21             

In [36]:
pd.options.display.max_rows = None

In [37]:
# Deriving Columns with loc()
orders_products_merged.loc[orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [38]:
orders_products_merged.loc[(orders_products_merged['max_order'] <= 40) & (orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [39]:
orders_products_merged.loc[orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [40]:
orders_products_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15877395
Loyal customer      10284396
New customer         6244250
Name: loyalty_flag, dtype: int64

In [41]:
orders_products_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  ...  aisle_id department_id  prices  _merge    price_range_loc  \
0        0.0  ...        77             7     9.0    both  Mid-range product   
1        1.0  ...        77             7     9.0    both  Mid-range product   
2        1.0  ...        77             7     9.0    both  Mid-range product   
3        1.0  ...        77             7     9.0    both  Mid-range product   
4        1.0  ...        77             7     9.0    both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy days        Average orders        10   
1      Least busy      Least busy days        Average orders        10   
2      Least busy      Least busy days           Most orders        10   
3      Least busy      Least busy days        Average orders        10   
4      Least busy      Least busy days           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 22 columns]

In [42]:
orders_products_merged[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

There is a greater range of department_id for entire dataframe in compared to the subset of dataframe of 1 million. 

With a greater range of department_id, the lowest mean value is department_id 5 with a mean value of 15.213779, while the highest mean value is department_id 21 with a mean value of 22.902379. 

With the subset of dataframe of 1 million, the lowest mean value is department_id is 17 with the mean value of 11.294069; the highest mean value is department_id is 16	with the mean value of 19.463012. 	


# 05. Check the spending habits

In [43]:
# Check the spending habits of three tpes of customers - check product prices for each loyalty category
orders_products_merged.groupby('loyalty_flag').agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386229  1.0  99999.0
New customer      13.294406  1.0  99999.0
Regular customer  12.495501  1.0  99999.0

# 06. Check the types of spenders

In [44]:
# Target customers based on spending, create a column of mean price by users

orders_products_merged['mean_order'] = orders_products_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [45]:
orders_products_merged.shape

(32406041, 23)

In [46]:
orders_products_merged.head(10)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   
5             5   3367565        1             6                   2   
6             6    550135        1             7                   1   
7             7   3108588        1             8                   1   
8             8   2295261        1             9                   1   
9             9   2550362        1            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   
5                  7                   19.0       196.0                1.0   
6                  9                   20.0       196.0                1.0   
7                 14                   14.0       196.0                2.0   
8                 16                    0.0       196.0                4.0   
9                  8                   30.0       196.0                1.0   

   reordered  ...  department_id prices  _merge    price_range_loc  \
0        0.0  ...              7    9.0    both  Mid-range product   
1        1.0  ...              7    9.0    both  Mid-range product   
2        1.0  ...              7    9.0    both  Mid-range product   
3        1.0  ...              7    9.0    both  Mid-range product   
4        1.0  ...              7    9.0    both  Mid-range product   
5        1.0  ...              7    9.0    both  Mid-range product   
6        1.0  ...              7    9.0    both  Mid-range product   
7        1.0  ...              7    9.0    both  Mid-range product   
8        1.0  ...              7    9.0    both  Mid-range product   
9        1.0  ...              7    9.0    both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy days        Average orders        10   
1      Least busy      Least busy days        Average orders        10   
2      Least busy      Least busy days           Most orders        10   
3      Least busy      Least busy days        Average orders        10   
4      Least busy      Least busy days           Most orders        10   
5  Regularly busy  Regularly busy days        Average orders        10   
6     Busiest day         Busiest days           Most orders        10   
7     Busiest day         Busiest days           Most orders        10   
8     Busiest day         Busiest days           Most orders        10   
9      Least busy      Least busy days        Average orders        10   

   loyalty_flag mean_order  
0  New customer   6.367797  
1  New customer   6.367797  
2  New customer   6.367797  
3  New customer   6.367797  
4  New customer   6.367797  
5  New customer   6.367797  
6  New customer   6.367797  
7  New customer   6.367797  
8  New customer   6.367797  
9  New customer   6.367797  

[10 rows x 23 columns]

In [47]:
# Categories spending flag based on mean price
orders_products_merged.loc[orders_products_merged['mean_order']<10, 'spending_flag'] = 'Low spender'

In [48]:
orders_products_merged.loc[orders_products_merged['mean_order']>=10, 'spending_flag'] = 'High spender'

In [49]:
orders_products_merged['spending_flag'].value_counts(dropna=False)

Low spender     31771789
High spender      634252
Name: spending_flag, dtype: int64

In [55]:
orders_products_merged[['user_id', 'mean_order', 'spending_flag']].head(100)

user_id  mean_order spending_flag
0         1    6.367797   Low spender
1         1    6.367797   Low spender
2         1    6.367797   Low spender
3         1    6.367797   Low spender
4         1    6.367797   Low spender
5         1    6.367797   Low spender
6         1    6.367797   Low spender
7         1    6.367797   Low spender
8         1    6.367797   Low spender
9         1    6.367797   Low spender
10       15    3.980556   Low spender
11       15    3.980556   Low spender
12       15    3.980556   Low spender
13       15    3.980556   Low spender
14       15    3.980556   Low spender
15       19    8.372059   Low spender
16       19    8.372059   Low spender
17       19    8.372059   Low spender
18       21    7.012683   Low spender
19       31    8.137793   Low spender
20       31    8.137793   Low spender
21       43    7.444898   Low spender
22       43    7.444898   Low spender
23       52    6.356287   Low spender
24       52    6.356287   Low spender
25       52    6.356287   Low spender
26       52    6.356287   Low spender
27       52    6.356287   Low spender
28       52    6.356287   Low spender
29       52    6.356287   Low spender
30       52    6.356287   Low spender
31       52    6.356287   Low spender
32       52    6.356287   Low spender
33       52    6.356287   Low spender
34       52    6.356287   Low spender
35       52    6.356287   Low spender
36       52    6.356287   Low spender
37       67    7.656790   Low spender
38       67    7.656790   Low spender
39       67    7.656790   Low spender
40       67    7.656790   Low spender
41       67    7.656790   Low spender
42       67    7.656790   Low spender
43       67    7.656790   Low spender
44       67    7.656790   Low spender
45       67    7.656790   Low spender
46       67    7.656790   Low spender
47       67    7.656790   Low spender
48       67    7.656790   Low spender
49       67    7.656790   Low spender
50       67    7.656790   Low spender
51       67    7.656790   Low spender
52       67    7.656790   Low spender
53       67    7.656790   Low spender
54       67    7.656790   Low spender
55       67    7.656790   Low spender
56       81    8.165385   Low spender
57       81    8.165385   Low spender
58       82    6.441221   Low spender
59       82    6.441221   Low spender
60       82    6.441221   Low spender
61       82    6.441221   Low spender
62       82    6.441221   Low spender
63       82    6.441221   Low spender
64       82    6.441221   Low spender
65       82    6.441221   Low spender
66       82    6.441221   Low spender
67       98    8.028000   Low spender
68       98    8.028000   Low spender
69       98    8.028000   Low spender
70       98    8.028000   Low spender
71       98    8.028000   Low spender
72       98    8.028000   Low spender
73       98    8.028000   Low spender
74      109    5.850725   Low spender
75      120    9.385714   Low spender
76      120    9.385714   Low spender
77      185    7.715000   Low spender
78      195    7.149730   Low spender
79      195    7.149730   Low spender
80      195    7.149730   Low spender
81      195    7.149730   Low spender
82      195    7.149730   Low spender
83      195    7.149730   Low spender
84      195    7.149730   Low spender
85      195    7.149730   Low spender
86      195    7.149730   Low spender
87      195    7.149730   Low spender
88      195    7.149730   Low spender
89      195    7.149730   Low spender
90      222    6.862397   Low spender
91      222    6.862397   Low spender
92      290    7.285597   Low spender
93      290    7.285597   Low spender
94      331    7.705128   Low spender
95      360   10.006250  High spender
96      377    8.496552   Low spender
97      387    7.396610   Low spender
98      420    7.387805   Low spender
99      420    7.387805   Low spender

# 07. Check the frequent vs non-frequent customers

In [53]:
# Create a frequency vs non-frequency customers based on median days since previous order, create a column of median days since previous orders by users

orders_products_merged['median_days_since_last_order'] = orders_products_merged.groupby(['user_id'])['days_since_last_order'].transform(np.mean)

In [54]:
orders_products_merged.shape

(32406041, 25)

In [56]:
orders_products_merged.head(10)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   
5             5   3367565        1             6                   2   
6             6    550135        1             7                   1   
7             7   3108588        1             8                   1   
8             8   2295261        1             9                   1   
9             9   2550362        1            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   
5                  7                   19.0       196.0                1.0   
6                  9                   20.0       196.0                1.0   
7                 14                   14.0       196.0                2.0   
8                 16                    0.0       196.0                4.0   
9                  8                   30.0       196.0                1.0   

   reordered  ...  _merge    price_range_loc     busiest_day  \
0        0.0  ...    both  Mid-range product  Regularly busy   
1        1.0  ...    both  Mid-range product      Least busy   
2        1.0  ...    both  Mid-range product      Least busy   
3        1.0  ...    both  Mid-range product      Least busy   
4        1.0  ...    both  Mid-range product      Least busy   
5        1.0  ...    both  Mid-range product  Regularly busy   
6        1.0  ...    both  Mid-range product     Busiest day   
7        1.0  ...    both  Mid-range product     Busiest day   
8        1.0  ...    both  Mid-range product     Busiest day   
9        1.0  ...    both  Mid-range product      Least busy   

          busiest_days  busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy days         Average orders        10  New customer   
1      Least busy days         Average orders        10  New customer   
2      Least busy days            Most orders        10  New customer   
3      Least busy days         Average orders        10  New customer   
4      Least busy days            Most orders        10  New customer   
5  Regularly busy days         Average orders        10  New customer   
6         Busiest days            Most orders        10  New customer   
7         Busiest days            Most orders        10  New customer   
8         Busiest days            Most orders        10  New customer   
9      Least busy days         Average orders        10  New customer   

  mean_order spending_flag median_days_since_last_order  
0   6.367797   Low spender                    20.259259  
1   6.367797   Low spender                    20.259259  
2   6.367797   Low spender                    20.259259  
3   6.367797   Low spender                    20.259259  
4   6.367797   Low spender                    20.259259  
5   6.367797   Low spender                    20.259259  
6   6.367797   Low spender                    20.259259  
7   6.367797   Low spender                    20.259259  
8   6.367797   Low spender                    20.259259  
9   6.367797   Low spender                    20.259259  

[10 rows x 25 columns]

In [65]:
# Categories frequent vs unfrequent users based on median days since last orders
orders_products_merged.loc[orders_products_merged['median_days_since_last_order']>20, 'Frequency_flag'] = 'Non-Frequent customer'

In [66]:
orders_products_merged.loc[(orders_products_merged['median_days_since_last_order']>10) & (orders_products_merged['median_days_since_last_order']<=20), 'Frequency_flag'] = 'Regular customer'

In [67]:
orders_products_merged.loc[orders_products_merged['median_days_since_last_order']<=10, 'Frequency_flag'] = 'Frequent customer'

In [68]:
orders_products_merged.shape

(32406041, 27)

In [69]:
orders_products_merged['Frequency_flag'].value_counts(dropna = False)

Frequent customer        16469114
Regular customer         12521787
Non-Frequent customer     3415135
NaN                             5
Name: Frequency_flag, dtype: int64

In [70]:
orders_products_merged[['user_id', 'median_days_since_last_order', 'Frequency_flag']].head(100)

user_id  median_days_since_last_order         Frequency_flag
0         1                     20.259259  Non-Frequent customer
1         1                     20.259259  Non-Frequent customer
2         1                     20.259259  Non-Frequent customer
3         1                     20.259259  Non-Frequent customer
4         1                     20.259259  Non-Frequent customer
5         1                     20.259259  Non-Frequent customer
6         1                     20.259259  Non-Frequent customer
7         1                     20.259259  Non-Frequent customer
8         1                     20.259259  Non-Frequent customer
9         1                     20.259259  Non-Frequent customer
10       15                     11.029412       Regular customer
11       15                     11.029412       Regular customer
12       15                     11.029412       Regular customer
13       15                     11.029412       Regular customer
14       15                     11.029412       Regular customer
15       19                      8.040698      Frequent customer
16       19                      8.040698      Frequent customer
17       19                      8.040698      Frequent customer
18       21                     10.055000       Regular customer
19       31                      6.073427      Frequent customer
20       31                      6.073427      Frequent customer
21       43                     10.549296       Regular customer
22       43                     10.549296       Regular customer
23       52                      9.155280      Frequent customer
24       52                      9.155280      Frequent customer
25       52                      9.155280      Frequent customer
26       52                      9.155280      Frequent customer
27       52                      9.155280      Frequent customer
28       52                      9.155280      Frequent customer
29       52                      9.155280      Frequent customer
30       52                      9.155280      Frequent customer
31       52                      9.155280      Frequent customer
32       52                      9.155280      Frequent customer
33       52                      9.155280      Frequent customer
34       52                      9.155280      Frequent customer
35       52                      9.155280      Frequent customer
36       52                      9.155280      Frequent customer
37       67                      7.714286      Frequent customer
38       67                      7.714286      Frequent customer
39       67                      7.714286      Frequent customer
40       67                      7.714286      Frequent customer
41       67                      7.714286      Frequent customer
42       67                      7.714286      Frequent customer
43       67                      7.714286      Frequent customer
44       67                      7.714286      Frequent customer
45       67                      7.714286      Frequent customer
46       67                      7.714286      Frequent customer
47       67                      7.714286      Frequent customer
48       67                      7.714286      Frequent customer
49       67                      7.714286      Frequent customer
50       67                      7.714286      Frequent customer
51       67                      7.714286      Frequent customer
52       67                      7.714286      Frequent customer
53       67                      7.714286      Frequent customer
54       67                      7.714286      Frequent customer
55       67                      7.714286      Frequent customer
56       81                     26.086957  Non-Frequent customer
57       81                     26.086957  Non-Frequent customer
58       82                     12.188525       Regular customer
59       82                     12.188525       Regular customer
60       82                     12.1885

# 08. Export File

In [ ]:
# Export new file as pkl
orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_rev_02.pkl'))